### PR TITLE
Obsolete - separate Verrazzano context from component context

### PR DESCRIPTION
**Pull request by pfmackin**
Obsolete

Verrazzano has a ComponentContext struct that is overloaded and used also as a Verrazzano CR context.  This is confusing and makes sections of the code hard to understand. This PR introduces a VerrazzanoComponent that applies to all components.  Also, refactor the ComponentContext initialization.

One thing to note is that we now create the effective CR at the controller level for the VerrazzanoContext vs the Component level in the ComponentContext.

# Checklist 

As the author of this PR, I have:

- [ ] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/spi/component.go [hunk 1]
```diff
@@ -20,46 +20,42 @@ type ComponentContext interface {
 	ActualCR() *vzapi.Verrazzano
 	// EffectiveCR returns the effective merged Verrazzano CR
 	EffectiveCR() *vzapi.Verrazzano
-	// IsDryRun indicates the component context is in DryRun mode
+	// IsDryRun indicates the componentName context is in DryRun mode
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/spi/component.go:23]: Is this change intentional?

File: platform-operator/controllers/verrazzano/component/spi/component_context.go [hunk 2]
```diff
@@ -5,46 +5,64 @@ package spi
...
-		client:      c,
-		dryRun:      dryRun,
-		cr:          actualCR,
-		effectiveCR: effectiveCR,
-	}, nil
+		componentName: compName,
+		log:           log,
+		client:        vzContext.Client,
+		dryRun:        vzContext.DryRun,
+		cr:            vzContext.ActualCR,
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/spi/component_context.go:54]: Would actualCR be more clear?

File: platform-operator/controllers/verrazzano/component/spi/component_context.go [hunk 3]
```diff
@@ -5,46 +5,64 @@ package spi
...
+	}
 }
 
 // NewFakeContext creates a fake ComponentContext for unit testing purposes
-// c Kubernetes client
-// actualCR The user-supplied Verrazzano CR
-// dryRun Dry-run indicator
-// profilesDir Optional override to the location of the profiles dir; if not provided, EffectiveCR == ActualCR
 func NewFakeContext(c clipkg.Client, actualCR *vzapi.Verrazzano, dryRun bool, profilesDir ...string) ComponentContext {
+	return NewFakeContextWithNameAndOperation(c, actualCR, dryRun, "", "", profilesDir...)
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/spi/component_context.go:61]: Should NewFakeContext be in this file?

File: platform-operator/controllers/verrazzano/component/spi/component.go (hunk 1 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/spi/component.go:23]: yeah, looks like a Goland refactoring victim.

File: platform-operator/controllers/verrazzano/component/spi/component.go [hunk 4]
```diff
@@ -20,46 +20,42 @@ type ComponentContext interface {
...
 	// Copy returns a copy of the current context
 	Copy() ComponentContext
-	// Init returns a copy of the current context with an updated logging component field
-	Init(comp string) ComponentContext
-	// Operation specifies the logging operation field
-	Operation(op string) ComponentContext
 	// GetOperation returns the operation object in the context
 	GetOperation() string
-	// GetComponent returns the component object in the context
+	// GetComponent returns the componentName object in the context
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/spi/component.go]: similar refactoring sideffect here?

File: platform-operator/controllers/verrazzano/component/spi/component_context.go (hunk 2 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/component/spi/component_context.go:54]: I can change that, i was using what was there.

File: platform-operator/controllers/verrazzano/component/spi/component_context_test.go [hunk 5]
```diff
@@ -152,8 +157,8 @@ func TestContextProfilesMerge(t *testing.T) {
 			assert.NotNil(context, "Context was nil")
 			assert.NotNil(context.ActualCR(), "Actual CR was nil")
 			assert.Equal(test.actualCR, *context.ActualCR(), "Actual CR unexpectedly modified")
+			context.EffectiveCR().Status = vzapi.VerrazzanoStatus{}
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/spi/component_context_test.go:160]: This should be after the next line.  Is this just to "fake out" the assert.Equal call below?

File: platform-operator/controllers/verrazzano/component/spi/component_context.go (hunk 3 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/component/spi/component_context.go:61]: I tried moving to a test file but doesn't compile.  It was already there

File: platform-operator/controllers/verrazzano/context/vzcontext.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	Client clipkg.Client
+	// dryRun If true, do a dry run of operations
+	DryRun bool
+	// ActualCR is the CR passed to top level Reconcile.  It epresents the desired Verrazzano state in the cluster
+	ActualCR *vzapi.Verrazzano
+	// effectiveCR Represents the configuration resulting from any named profiles used and any configured overrides in the CR
+	EffectiveCR *vzapi.Verrazzano
+}
+
+// NewContext creates a ComponentContext from a raw CR
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/context/vzcontext.go:34]: Comment doesn't match function name.  Is New as a function name a good idea?

File: platform-operator/controllers/verrazzano/context/vzcontext.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+func New(log vzlog.VerrazzanoLogger, c clipkg.Client, actualCR *vzapi.Verrazzano, dryRun bool) (VerrazzanoContext, error) {
+	// Generate the effective CR based ond the declared profile and any overrides in the user-supplied one
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/context/vzcontext.go:36]: typo

File: platform-operator/controllers/verrazzano/controller.go [hunk 8]
```diff
@@ -138,7 +140,7 @@ func (r *Reconciler) doReconcile(log vzlog.VerrazzanoLogger, vz *installv1alpha1
 		return reconcile.Result{Requeue: true}, nil
 	}
 
-	spiContext, err := spi.NewContext(log, r, vz, r.DryRun)
+	vzContext, err := vzctx.New(log, r, vz, r.DryRun)
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/controller.go:143]: Should this return a pointer?

File: platform-operator/controllers/verrazzano/component/spi/component_context_test.go (hunk 5 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/component/spi/component_context_test.go:160]: It is needed since we now clone the status from the actual CR to the effective CR, since we are now creating the effective CR earlier.

File: platform-operator/controllers/verrazzano/context/vzcontext.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+		profileFiles = append(profileFiles, config.GetProfile(profile))
+	}
+	// Merge the profile files into an effective profile YAML string
+	effectiveCR, err := transform.MergeProfiles(actualCR, profileFiles...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Copy actual status to effective CR
+	actualCR.Status.DeepCopyInto(&effectiveCR.Status)
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/context/vzcontext.go:73]: Originally, I'd left the effective CR status empty, so it didn't duplicate what was in the actualCR, but maybe that was more confusing.

I think the status field should be populated via the merge above anyway, if we truly want to duplicate the status field in the Effective CR.  I think initially, we had thought we only wanted the Spec field, then we decided to keep the Metadata and other info.  But not 100% sure we want to have the status duplicated.

File: platform-operator/controllers/verrazzano/context/vzcontext.go (hunk 6 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/context/vzcontext.go:34]: New is a go convention if you have single purpose package like this.

File: platform-operator/controllers/verrazzano/component/spi/component_context.go (hunk 3 above)
Comment (kminder) [reply platform-operator/controllers/verrazzano/component/spi/component_context.go:61]: ok

File: platform-operator/controllers/verrazzano/controller.go [hunk 10]
```diff
@@ -688,7 +689,8 @@ func (r *Reconciler) checkComponentReadyState(context spi.ComponentContext) (boo
 
 	// Return false if any enabled component is not ready
 	for _, comp := range registry.GetComponents() {
-		if comp.IsEnabled(context) && cr.Status.Components[comp.Name()].State != installv1alpha1.CompStateReady {
+		compContext := spi.NewComponentContext(vzContext, comp.Name(), "")
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/controller.go:692]: Wonder if we need create an operation name here instead of ""

File: platform-operator/controllers/verrazzano/component/spi/component_context.go (hunk 3 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/spi/component_context.go:61]: I had originally tried doing that, but then it couldn't be shared with other test code.  I thought it lined up with things like the fake controllerruntime client, although perhaps this kind of stuff should be in its own `fake` package?

**Review by kminder — COMMENTED** [review]
Approved

**Review by kminder — DISMISSED** [review]
Approved.

File: platform-operator/controllers/verrazzano/context/vzcontext.go (hunk 9 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/context/vzcontext.go:74]: This is needed for upgrade.  We can look at removing the code that checks the status now that the failed state is not used.

File: platform-operator/controllers/verrazzano/controller.go (hunk 10 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/controller.go:692]: no really, this is logging at the global level

File: platform-operator/controllers/verrazzano/component/spi/component.go (hunk 4 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/component/spi/component.go]: fixed them all

File: platform-operator/controllers/verrazzano/context/vzcontext.go (hunk 7 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/context/vzcontext.go]: I also changed this to NewVerrazzanoContext

**Review by kminder — APPROVED** [review]
Approved.

**Review by michael-cico — APPROVED** [review]
